flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java

```diff
@@ -20,59 +20,118 @@
+
+		assertThat(checkpointCoordinator, Matchers.notNullValue());
+		assertThat(checkpointCoordinator.isShutdown(), is(false));
+
 		graph.failGlobal(new Exception("Test Exception"));
 
-		verify(counter, times(1)).shutdown(JobStatus.FAILED);
-		verify(store, times(1)).shutdown(eq(JobStatus.FAILED));
+		assertThat(checkpointCoordinator.isShutdown(), is(true));
+		assertThat(counterShutdownFuture.get(), is(JobStatus.FAILED));
+		assertThat(storeShutdownFuture.get(), is(JobStatus.FAILED));
 	}
 
 	/**
-	 * Tests that a suspended checkpoint coordinator calls suspend on
-	 * the store and counter.
+	 * Tests that the checkpoint coordinator is shut down if the execution graph
+	 * is suspended.
 	 */
 	@Test
-	public void testSuspendCheckpointCoordinator() throws Exception {
-		CheckpointIDCounter counter = mock(CheckpointIDCounter.class);
-		CompletedCheckpointStore store = mock(CompletedCheckpointStore.class);
+	public void testShutdownCheckpointCoordinatorOnSuspend() throws Exception {
+		final CompletableFuture<JobStatus> counterShutdownFuture = new CompletableFuture<>();
+		CheckpointIDCounter counter = new TestingCheckpointIDCounter(counterShutdownFuture);
+
+		final CompletableFuture<JobStatus> storeShutdownFuture = new CompletableFuture<>();
+		CompletedCheckpointStore store = new TestingCompletedCheckpointStore(storeShutdownFuture);
 
 		ExecutionGraph graph = createExecutionGraphAndEnableCheckpointing(counter, store);
+		final CheckpointCoordinator checkpointCoordinator = graph.getCheckpointCoordinator();
+
+		assertThat(checkpointCoordinator, Matchers.notNullValue());
+		assertThat(checkpointCoordinator.isShutdown(), is(false));
+
 		graph.suspend(new Exception("Test Exception"));
 
-		// No shutdown
-		verify(counter, times(1)).shutdown(eq(JobStatus.SUSPENDED));
-		verify(store, times(1)).shutdown(eq(JobStatus.SUSPENDED));
+		assertThat(checkpointCoordinator.isShutdown(), is(true));
+		assertThat(counterShutdownFuture.get(), is(JobStatus.SUSPENDED));
+		assertThat(storeShutdownFuture.get(), is(JobStatus.SUSPENDED));
+	}
+
+	/**
+	 * Tests that the checkpoint coordinator is shut down if the execution graph
+	 * is finished.
+	 */
+	@Test
+	public void testShutdownCheckpointCoordinatorOnFinished() throws Exception {
+		final CompletableFuture<JobStatus> counterShutdownFuture = new CompletableFuture<>();
+		CheckpointIDCounter counter = new TestingCheckpointIDCounter(counterShutdownFuture);
+
+		final CompletableFuture<JobStatus> storeShutdownFuture = new CompletableFuture<>();
+		CompletedCheckpointStore store = new TestingCompletedCheckpointStore(storeShutdownFuture);
+
+		ExecutionGraph graph = createExecutionGraphAndEnableCheckpointing(counter, store);
+		final CheckpointCoordinator checkpointCoordinator = graph.getCheckpointCoordinator();
+
+		assertThat(checkpointCoordinator, Matchers.notNullValue());
+		assertThat(checkpointCoordinator.isShutdown(), is(false));
+
+		graph.scheduleForExecution();
```

Comment:
Kind of unrelated: Where have we documented what state transitions are valid?

Reply:
https://ci.apache.org/projects/flink/flink-docs-stable/internals/job_scheduling.html but this might actually be partially out dated.